### PR TITLE
8331861: [PPC64] Implement load / store assembler functions which take an Address object

### PR DESCRIPTION
**Pull request by varada1110**
Load and store assembly instructions which takes Address object as argument. 

Tier 1 testing successful on linux-ppc64le and aix-ppc (fastdebug level)

JBS : [JDK-8331861](https://bugs.openjdk.org/browse/JDK-8331861)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331861](https://bugs.openjdk.org/browse/JDK-8331861): [PPC64] Implement load / store assembler functions which take an Address object (**Enhancement** - P4)

### Reviewers
 * [Amit Kumar](https://openjdk.org/census#amitkumar) (@offamitkumar - Committer)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21492/head:pull/21492` \
`$ git checkout pull/21492`

Update a local copy of the PR: \
`$ git checkout pull/21492` \
`$ git pull https://git.openjdk.org/jdk.git pull/21492/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21492`

View PR using the GUI difftool: \
`$ git pr show -t 21492`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21492.diff">https://git.openjdk.org/jdk/pull/21492.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21492#issuecomment-2442095275)

### CODE REVIEW

**Review by offamitkumar — CHANGES_REQUESTED** [review]
Can you apply this patch, I did the change with regex, so Please verify before pushing it. 

```diff
diff --git a/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp b/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
index 684c06614a9..f9207d89615 100644
--- a/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
+++ b/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
@@ -157,9 +157,9 @@ void LIR_Assembler::osr_entry() {
               mo = frame_map()->address_for_monitor_object(i);
       assert(ml.index() == noreg && mo.index() == noreg, "sanity");
       __ ld(R0, slot_offset + 0, OSR_buf);
-      __ std(R0, ml.disp(), ml.base());
+      __ std(R0, ml);
       __ ld(R0, slot_offset + 1*BytesPerWord, OSR_buf);
-      __ std(R0, mo.disp(), mo.base());
+      __ std(R0, mo);
     }
   }
 }
@@ -581,14 +581,14 @@ void LIR_Assembler::emit_opConvert(LIR_OpConvert* op) {
       __ fcmpu(CCR0, rsrc, rsrc);
       if (dst_in_memory) {
         __ li(R0, 0); // 0 in case of NAN
-        __ std(R0, addr.disp(), addr.base());
+        __ std(R0, addr);
       } else {
         __ li(dst->as_register(), 0);
       }
       __ bso(CCR0, L);
       __ fctiwz(rsrc, rsrc); // USE_KILL
       if (dst_in_memory) {
-        __ stfd(rsrc, addr.disp(), addr.base());
+        __ stfd(rsrc, addr);
       } else {
         __ mffprd(dst->as_register(), rsrc);
       }
@@ -605,14 +605,14 @@ void LIR_Assembler::emit_opConvert(LIR_OpConvert* op) {
       __ fcmpu(CCR0, rsrc, rsrc);
       if (dst_in_memory) {
         __ li(R0, 0); // 0 in case of NAN
-        __ std(R0, addr.disp(), addr.base());
+        __ std(R0, addr);
       } else {
         __ li(dst->as_register_lo(), 0);
       }
       __ bso(CCR0, L);
       __ fctidz(rsrc, rsrc); // USE_KILL
       if (dst_in_memory) {
-        __ stfd(rsrc, addr.disp(), addr.base());
+        __ stfd(rsrc, addr);
       } else {
         __ mffprd(dst->as_register_lo(), rsrc);
       }
@@ -873,20 +873,20 @@ void LIR_Assembler::const2stack(LIR_Opr src, LIR_Opr dest) {
       int value = c->as_jint_bits();
       __ load_const_optimized(src_reg, value);
       Address addr = frame_map()->address_for_slot(dest->single_stack_ix());
-      __ stw(src_reg, addr.disp(), addr.base());
+      __ stw(src_reg, addr);
       break;
     }
     case T_ADDRESS: {
       int value = c->as_jint_bits();
       __ load_const_optimized(src_reg, value);
       Address addr = frame_map()->address_for_slot(dest->single_stack_ix());
-      __ std(src_reg, addr.disp(), addr.base());
+      __ std(src_reg, addr);
       break;
     }
     case T_OBJECT: {
       jobject2reg(c->as_jobject(), src_reg);
       Address addr = frame_map()->address_for_slot(dest->single_stack_ix());
-      __ std(src_reg, addr.disp(), addr.base());
+      __ std(src_reg, addr);
       break;
     }
     case T_LONG:
@@ -894,7 +894,7 @@ void LIR_Assembler::const2stack(LIR_Opr src, LIR_Opr dest) {
       int value = c->as_jlong_bits();
       __ load_const_optimized(src_reg, value);
       Address addr = frame_map()->address_for_double_slot(dest->double_stack_ix());
-      __ std(src_reg, addr.disp(), addr.base());
+      __ std(src_reg, addr);
       break;
     }
     default:
@@ -1070,24 +1070,24 @@ void LIR_Assembler::stack2stack(LIR_Opr src, LIR_Opr dest, BasicType type) {
     case T_FLOAT: {
       Address from = frame_map()->address_for_slot(src->single_stack_ix());
       Address to   = frame_map()->address_for_slot(dest->single_stack_ix());
-      __ lwz(tmp, from.disp(), from.base());
-      __ stw(tmp, to.disp(), to.base());
+      __ lwz(tmp, from);
+      __ stw(tmp, to);
       break;
     }
     case T_ADDRESS:
     case T_OBJECT: {
       Address from = frame_map()->address_for_slot(src->single_stack_ix());
       Address to   = frame_map()->address_for_slot(dest->single_stack_ix());
-      __ ld(tmp, from.disp(), from.base());
-      __ std(tmp, to.disp(), to.base());
+      __ ld(tmp, from);
+      __ std(tmp, to);
       break;
     }
     case T_LONG:
     case T_DOUBLE: {
       Address from = frame_map()->address_for_double_slot(src->double_stack_ix());
       Address to   = frame_map()->address_for_double_slot(dest->double_stack_ix());
-      __ ld(tmp, from.disp(), from.base());
-      __ std(tmp, to.disp(), to.base());
+      __ ld(tmp, from);
+      __ std(tmp, to);
       break;
     }
 
diff --git a/src/hotspot/cpu/ppc/gc/z/zBarrierSetAssembler_ppc.cpp b/src/hotspot/cpu/ppc/gc/z/zBarrierSetAssembler_ppc.cpp
index 89ab1b1edee..0307c60087f 100644
--- a/src/hotspot/cpu/ppc/gc/z/zBarrierSetAssembler_ppc.cpp
+++ b/src/hotspot/cpu/ppc/gc/z/zBarrierSetAssembler_ppc.cpp
@@ -610,14 +610,14 @@ void ZBarrierSetAssembler::try_resolve_jobject_in_native(MacroAssembler* masm, R
 
   // Resolve global handle
   __ ld(dst, 0, dst);
-  __ ld(tmp, load_bad_mask.disp(), load_bad_mask.base());
+  __ ld(tmp, load_bad_mask);
   __ b(check_color);
 
   __ bind(weak_tagged);
 
   // Resolve weak handle
   __ ld(dst, 0, dst);
-  __ ld(tmp, mark_bad_mask.disp(), mark_bad_mask.base());
+  __ ld(tmp, mark_bad_mask);
 
   __ bind(check_color);
   __ and_(tmp, tmp, dst);
```

**Comment by TheRealMDoerr** [comment]
Please don't forget that an `Address` either has `_index` or `_disp` (but not both: https://github.com/openjdk/jdk/blob/f56a154132f7e66b1b65adfa2aa937119999b14a/src/hotspot/cpu/ppc/assembler_ppc.hpp#L44).

**Comment by TheRealMDoerr** [comment]
Idea: Use the `RegisterOrConstant` version. This covers all cases including large offset and index. E.g.
`inline void stw(  Register d, Address &a, Register tmp = noreg);`
```
inline void Assembler::stw(  Register d, Address &a, Register tmp) {
  stw(d, a.index() != noreg ? RegisterOrConstant(a.index()) : RegisterOrConstant(a.disp()), a.base(), tmp);
}
```
I'd move them into a separate section.

**Comment by varada1110** [comment]
Tier1 testing on linux-ppc64le has completed successfully. I am now running tests for aix-ppc

File: src/hotspot/cpu/ppc/assembler_ppc.hpp [hunk 1]
```diff
@@ -2522,6 +2522,17 @@ class Assembler : public AbstractAssembler {
...
+  void ld( Register d, Address &a);
+  void lwa( Register d, Address &a);
+  void lwz( Register d, Address &a);
+  void lha( Register d, Address &a);
+  void lhz( Register d, Address &a);
+  void lbz( Register d, Address &a);
+  void std( Register d, Address &a, Register tmp = noreg);
+  void stw( Register d, Address &a, Register tmp = noreg);
+  void sth( Register d, Address &a, Register tmp = noreg);
+  void stb( Register d, Address &a, Register tmp = noreg);
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/assembler_ppc.hpp:2534]: Spaces before "Register d" are uncommon. Please remove them (you can keep the one for ld to align it with the other ones).

File: src/hotspot/cpu/ppc/assembler_ppc.inline.hpp [hunk 2]
```diff
@@ -371,19 +377,23 @@ inline void Assembler::ld_ptr(Register d, ByteSize b, Register s1) { ld(d, in_by
 
 //  PPC 1, section 3.3.3 Fixed-Point Store Instructions
 inline void Assembler::stwx( Register d, Register s1, Register s2) { emit_int32(STWX_OPCODE | rs(d) | ra0mem(s1) | rb(s2));}
+inline void Assembler::stw(  Register d, Address &a, Register tmp) { stw(d, a.index() != noreg ? RegisterOrConstant(a.index()) : RegisterOrConstant(a.disp()), a.base(), tmp); }
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/assembler_ppc.inline.hpp:380]: The line is very long. Better break it to:
```
inline void Assembler::stw(  Register d, Address &a, Register tmp) {
  stw(d, a.index() != noreg ? RegisterOrConstant(a.index()) : RegisterOrConstant(a.disp()), a.base(), tmp);
}
```

File: src/hotspot/cpu/ppc/assembler_ppc.inline.hpp [hunk 3]
```diff
@@ -338,28 +338,34 @@ inline void Assembler::insrwi(  Register a, Register s, int n,   int b)
 
 // PPC 1, section 3.3.2 Fixed-Point Load Instructions
 inline void Assembler::lwzx( Register d, Register s1, Register s2) { emit_int32(LWZX_OPCODE | rt(d) | ra0mem(s1) | rb(s2));}
+inline void Assembler::lwz(  Register d, Address &a) { lwz(d, a.disp(), a.base()); }
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/assembler_ppc.inline.hpp:341]: The load instructions need the same adaptation as the store instructions (just without the tmp Register).
```
inline void Assembler::lwz(  Register d, Address &a) {
  lwz(d, a.index() != noreg ? RegisterOrConstant(a.index()) : RegisterOrConstant(a.disp()), a.base());
}
```

**Review by TheRealMDoerr — COMMENTED** [review]
Thanks this looks better.

File: src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp [hunk 4]
```diff
@@ -1070,23 +1070,23 @@ void LIR_Assembler::stack2stack(LIR_Opr src, LIR_Opr dest, BasicType type) {
...
       __ std(tmp, to.disp(), to.base());
       break;
     }
     case T_LONG:
     case T_DOUBLE: {
       Address from = frame_map()->address_for_double_slot(src->double_stack_ix());
       Address to   = frame_map()->address_for_double_slot(dest->double_stack_ix());
-      __ ld(tmp, from.disp(), from.base());
+      __ ld(tmp, from);
       __ std(tmp, to.disp(), to.base());
```
Comment (offamitkumar) [inline src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp:1090]: why these store instruction left behind ? 

Can't we just pass address and give it scratch register to play with ?

Comment (offamitkumar) [reply src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp:1090]: Oh, we don't even need `tmp` and it defaults to `noreg`. Even better.

File: src/hotspot/cpu/ppc/assembler_ppc.hpp (hunk 1 above)
Comment (varada1110) [reply src/hotspot/cpu/ppc/assembler_ppc.hpp]: Thank you Martin. I have fixed the alignment for both the cases (RegisterOrConstant and Address obj). Also I have adapted the same changes to load instructions.

File: src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp (hunk 4 above)
Comment (varada1110) [reply src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp]: Thanks for suggestion! I have adapted the changes for store instructions. Testing (tier1) on linux-ppc64le is successfully completed. Testing to be done for aix-ppc.

File: src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp [hunk 5]
```diff
@@ -157,9 +157,9 @@ void LIR_Assembler::osr_entry() {
               mo = frame_map()->address_for_monitor_object(i);
       assert(ml.index() == noreg && mo.index() == noreg, "sanity");
       __ ld(R0, slot_offset + 0, OSR_buf);
-      __ std(R0, ml.disp(), ml.base());
+      __ std(R0, ml, noreg);
```
Comment (offamitkumar) [inline src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp:160]: No, `tmp=noreg` is default case.

So 
```suggestion
      __ std(R0, ml);
```
should work.

**Comment by offamitkumar** [comment]
@TheRealMDoerr are floating-point load/store instructions out of scope for this PR? 

I see couple of use cases: 
```cpp
./c1_LIRAssembler_ppc.cpp:591:        __ stfd(rsrc, addr.disp(), addr.base());
./c1_LIRAssembler_ppc.cpp:615:        __ stfd(rsrc, addr.disp(), addr.base());
```

**Review by offamitkumar — APPROVED** [review]
Looks Good

**Comment by TheRealMDoerr** [comment]
> @TheRealMDoerr are floating-point load/store instructions out of scope for this PR?
> 
> I see couple of use cases:
> 
> ```c++
> ./c1_LIRAssembler_ppc.cpp:591:        __ stfd(rsrc, addr.disp(), addr.base());
> ./c1_LIRAssembler_ppc.cpp:615:        __ stfd(rsrc, addr.disp(), addr.base());
> ```

That could be done, too, but floating point instructions are so rarely used, that we could skip them.

**Comment by varada1110** [comment]
Tier1 testing on aix-ppc has completed successfully.

**Review by TheRealMDoerr — APPROVED** [review]
LGTM.

**Comment by varada1110** [comment]
Thanks @TheRealMDoerr @offamitkumar 
/integrate

**Pull request closed without merging**